src/main/java/com/iota/iri/network/NeighborRouter.java

```diff
@@ -0,0 +1,962 @@
+            TransactionProcessingPipeline txPipeline) {
+        this.txRequester = txRequester;
+        this.txPipeline = txPipeline;
+        this.networkConfig = networkConfig;
+        this.protocolConfig = protocolConfig;
+
+        // reduce the coordinator address to its byte encoded representation
+        byte[] tritsEncodedCooAddress = new byte[protocolConfig.getCoordinator().toString().length()
+                * Converter.NUMBER_OF_TRITS_IN_A_TRYTE];
+        Converter.trits(protocolConfig.getCoordinator().toString(), tritsEncodedCooAddress, 0);
+        byteEncodedCooAddress = new byte[Handshake.BYTE_ENCODED_COO_ADDRESS_BYTES_LENGTH];
+        Converter.bytes(tritsEncodedCooAddress, byteEncodedCooAddress);
+    }
+
+    private void initNeighbors() {
+        // parse URIs
+        networkConfig.getNeighbors().stream()
+                .distinct()
+                .map(NeighborRouter::parseURI)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .forEach(uri -> reconnectPool.add(uri));
+    }
+
+    /**
+     * Starts a dedicated thread for the {@link NeighborRouter} and then starts the routing mechanism.
+     */
+    public void start() {
+        executor.execute(this::route);
+    }
+
+    /**
+     * <p>
+     * Starts the routing mechanism which first initialises connections to neighbors from the configuration and then
+     * continuously reads and writes messages from/to neighbors.
+     * </p>
+     * <p>
+     * This method will also try to reconnect to wanted neighbors by the given
+     * {@link BaseIotaConfig#getReconnectAttemptIntervalSeconds()} value.
+     * </p>
+     */
+    public void route() {
+        log.info("starting neighbor router");
+
+        // run selector loop
+        try {
+            selector = Selector.open();
+            serverSocketChannel = ServerSocketChannel.open();
+            serverSocketChannel.configureBlocking(false);
+            serverSocketChannel.register(selector, SelectionKey.OP_ACCEPT);
+            InetSocketAddress tcpBindAddr = new InetSocketAddress(networkConfig.getNeighboringSocketAddress(),
+                    networkConfig.getNeighboringSocketPort());
+            serverSocketChannel.socket().bind(tcpBindAddr);
+            log.info("bound server TCP socket to {}", tcpBindAddr);
+
+            // parse neighbors from configuration
+            initNeighbors();
+
+            // init connections to the wanted neighbors,
+            // this also ensures that the whitelists are updated with the corresponding
+            // IP addresses and domain name mappings.
+            connectToWantedNeighbors();
+
+            long lastReconnectAttempts = System.currentTimeMillis();
+            long reconnectAttemptTimeout = TimeUnit.SECONDS
+                    .toMillis(networkConfig.getReconnectAttemptIntervalSeconds());
+
+            while (!shutdown.get()) {
```

Comment:
You can also do `Thread.currentThread.isInterrupted()` here and save a variable

Reply:
Changed